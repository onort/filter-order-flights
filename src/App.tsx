/* tslint:disable no-console */
import React, { Component } from "react"
import { connect } from "react-redux"
import { Dispatch } from "redux"

import styles from "./App.module.scss"
import { fetchData, FetchDataAction } from "./redux/actions"
import { ApplicationState } from "./redux/reducers"
import { FilteringOptions, Itinerary } from "./types"
import {
  Filters,
  Footer,
  Header,
  OrderResults,
  Results,
  SearchBar,
  Pagination
} from "./components"
import data from "./data/mockData.json"
import { formatData } from "./utils"

interface State {
  activeIndex: number
  totalPages: number
}

interface PropsFromDispatch {
  fetchData: typeof fetchData
}

interface PropsFromState {
  filteringOptions: FilteringOptions
  itineraries: Itinerary[]
}

type Props = PropsFromDispatch & PropsFromState

class App extends Component<Props, State> {
  public itemsPerPage: number = 10

  public state = {
    activeIndex: 1,
    totalPages: Math.ceil(this.props.itineraries.length / this.itemsPerPage)
  }

  public componentDidMount() {
    this.props.fetchData()
  }

  public componentDidUpdate(prevProps: Props) {
    if (prevProps.itineraries.length !== this.props.itineraries.length) {
      this.setState({
        totalPages: Math.ceil(this.props.itineraries.length / this.itemsPerPage)
      })
    }
  }

  public handlePagintionClick = (page: number) => {
    this.setState({ activeIndex: page })
    // TO DECIDE: Scroll to ResultsTable instead of 0
    window.scrollTo(0, 0)
  }

  public render() {
    // console.log(data.result.Itineraries.find(d => d.Filter.Carriers.length > 1))
    // console.log(
    //   data.result.Itineraries.find(i => i.OutboundLegId.Segments.length > 1)
    // )
    // console.log("Format Data", formatData(data.result.Itineraries))
    const { activeIndex, totalPages } = this.state
    const dataToShow = this.props.itineraries.slice(
      (activeIndex - 1) * this.itemsPerPage,
      activeIndex * this.itemsPerPage
    )
    return (
      <div className={styles.app}>
        <Header />
        <SearchBar type="primary" />
        <OrderResults flightCount={25} />
        {this.props.itineraries.length > 0 && (
          <div className={styles.main}>
            <Filters />
            <Results data={dataToShow} />
            <Pagination
              activeIndex={activeIndex}
              totalPages={totalPages}
              onClick={this.handlePagintionClick}
            />
          </div>
        )}
        <SearchBar type="secondary" />
        <Footer />
      </div>
    )
  }
}

const mapStateToProps = ({
  filteringOptions,
  itineraries
}: ApplicationState) => ({
  filteringOptions,
  itineraries
})

const mapDispatchToProps = (dispatch: Dispatch<FetchDataAction>) => ({
  fetchData: () => dispatch(fetchData())
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(App)
