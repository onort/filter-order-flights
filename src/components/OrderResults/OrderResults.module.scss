@import "../../styles/main.scss";

.container {
  display: flex;
  max-width: 1120px;
  align-items: center;
  justify-content: space-between;
  margin: map-get($spacers, x-large) auto map-get($spacers, large);
}

.flightCount {
  display: block;
  color: $black;
  font-family: $font-family-heading;
  font-size: map-get($heading-sizes, h2);
  font-weight: 600;
}
