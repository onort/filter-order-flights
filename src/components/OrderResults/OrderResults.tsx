import React from "react"

import styles from "./OrderResults.module.scss"
import { Dropdown } from "../"

interface Props {
  className?: string
  flightCount: number
}

const OrderResults: React.SFC<Props> = props => {
  const { flightCount } = props
  return (
    <section className={styles.container}>
      {flightCount > 0 && (
        <span className={styles.flightCount}>
          {flightCount} Uçuş Listeleniyor
        </span>
      )}
      <Dropdown title="Sıralama" selectedItem="Option A" />
    </section>
  )
}

export default OrderResults
