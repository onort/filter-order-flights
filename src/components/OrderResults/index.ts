import OrderResults from "./OrderResults"

export default OrderResults
