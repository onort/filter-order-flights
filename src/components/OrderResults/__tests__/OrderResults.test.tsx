import React from "react"
import { mount, shallow } from "enzyme"

import OrderResults from "../"
import { Dropdown } from "../../"
import { orderingOptions } from "../../../App"

const mockOnSelect = jest.fn()

describe("<OrderResults />", () => {
  it("should match snapshot", () => {
    const wrapper = shallow(
      <OrderResults
        flightCount={10}
        onOrderSelect={mockOnSelect}
        orderingOptions={orderingOptions}
        selectedOrder={orderingOptions[0]}
      />
    )
    expect(wrapper).toMatchSnapshot()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it("should show correct flight amount", () => {
    const flightCount = 10
    const wrapper = shallow(
      <OrderResults
        flightCount={flightCount}
        onOrderSelect={mockOnSelect}
        orderingOptions={orderingOptions}
        selectedOrder={orderingOptions[0]}
      />
    )
    expect(wrapper.find(".flightCount").text()).toBe(
      `${flightCount} Uçuş Listeleniyor`
    )
  })

  it("should correct text if there are no flights", () => {
    const wrapper = shallow(
      <OrderResults
        flightCount={0}
        onOrderSelect={mockOnSelect}
        orderingOptions={orderingOptions}
        selectedOrder={orderingOptions[0]}
      />
    )
    expect(wrapper.find(".flightCount").text()).toBe(
      "Kriterlere Uygun Uçuş Bulunamadı"
    )
  })

  it("should have a Dropdown component as child", () => {
    const wrapper = mount(
      <OrderResults
        flightCount={0}
        onOrderSelect={mockOnSelect}
        orderingOptions={orderingOptions}
        selectedOrder={orderingOptions[0]}
      />
    )
    expect(wrapper.find(Dropdown).exists()).toBe(true)
    wrapper.unmount()
  })
})
