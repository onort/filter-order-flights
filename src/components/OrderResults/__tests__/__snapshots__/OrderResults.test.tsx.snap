// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<OrderResults /> should match snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <OrderResults
    flightCount={10}
    onOrderSelect={[MockFunction]}
    orderingOptions={
      Array [
        Object {
          "text": "Ucuzdan Pahalıya",
          "value": "minPrice",
        },
        Object {
          "text": "Pahalıdan Ucuza",
          "value": "maxPrice",
        },
        Object {
          "text": "Tarihe Göre Yakın",
          "value": "closest",
        },
        Object {
          "text": "Tarihe Göre Uzak",
          "value": "farthest",
        },
      ]
    }
    selectedOrder={
      Object {
        "text": "Ucuzdan Pahalıya",
        "value": "minPrice",
      }
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <span
          className="flightCount"
        >
          10
           Uçuş Listeleniyor
        </span>,
        <div
          className="buttonsContainer"
        >
          <Button
            className="button"
            primary={true}
            text="Yeniden Ara"
            type="button"
          />
          <Button
            className="button"
            primary={true}
            text="Filtrele"
            type="button"
          />
        </div>,
        <Dropdown
          onSelect={[MockFunction]}
          orderingOtions={
            Array [
              Object {
                "text": "Ucuzdan Pahalıya",
                "value": "minPrice",
              },
              Object {
                "text": "Pahalıdan Ucuza",
                "value": "maxPrice",
              },
              Object {
                "text": "Tarihe Göre Yakın",
                "value": "closest",
              },
              Object {
                "text": "Tarihe Göre Uzak",
                "value": "farthest",
              },
            ]
          }
          selectedOrder={
            Object {
              "text": "Ucuzdan Pahalıya",
              "value": "minPrice",
            }
          }
          title="Sıralama"
        />,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            10,
            " Uçuş Listeleniyor",
          ],
          "className": "flightCount",
        },
        "ref": null,
        "rendered": Array [
          10,
          " Uçuş Listeleniyor",
        ],
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <Button
              className="button"
              primary={true}
              text="Yeniden Ara"
              type="button"
            />,
            <Button
              className="button"
              primary={true}
              text="Filtrele"
              type="button"
            />,
          ],
          "className": "buttonsContainer",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "button",
              "primary": true,
              "text": "Yeniden Ara",
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "className": "button",
              "primary": true,
              "text": "Filtrele",
              "type": "button",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "onSelect": [MockFunction],
          "orderingOtions": Array [
            Object {
              "text": "Ucuzdan Pahalıya",
              "value": "minPrice",
            },
            Object {
              "text": "Pahalıdan Ucuza",
              "value": "maxPrice",
            },
            Object {
              "text": "Tarihe Göre Yakın",
              "value": "closest",
            },
            Object {
              "text": "Tarihe Göre Uzak",
              "value": "farthest",
            },
          ],
          "selectedOrder": Object {
            "text": "Ucuzdan Pahalıya",
            "value": "minPrice",
          },
          "title": "Sıralama",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <span
            className="flightCount"
          >
            10
             Uçuş Listeleniyor
          </span>,
          <div
            className="buttonsContainer"
          >
            <Button
              className="button"
              primary={true}
              text="Yeniden Ara"
              type="button"
            />
            <Button
              className="button"
              primary={true}
              text="Filtrele"
              type="button"
            />
          </div>,
          <Dropdown
            onSelect={[MockFunction]}
            orderingOtions={
              Array [
                Object {
                  "text": "Ucuzdan Pahalıya",
                  "value": "minPrice",
                },
                Object {
                  "text": "Pahalıdan Ucuza",
                  "value": "maxPrice",
                },
                Object {
                  "text": "Tarihe Göre Yakın",
                  "value": "closest",
                },
                Object {
                  "text": "Tarihe Göre Uzak",
                  "value": "farthest",
                },
              ]
            }
            selectedOrder={
              Object {
                "text": "Ucuzdan Pahalıya",
                "value": "minPrice",
              }
            }
            title="Sıralama"
          />,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              10,
              " Uçuş Listeleniyor",
            ],
            "className": "flightCount",
          },
          "ref": null,
          "rendered": Array [
            10,
            " Uçuş Listeleniyor",
          ],
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <Button
                className="button"
                primary={true}
                text="Yeniden Ara"
                type="button"
              />,
              <Button
                className="button"
                primary={true}
                text="Filtrele"
                type="button"
              />,
            ],
            "className": "buttonsContainer",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "button",
                "primary": true,
                "text": "Yeniden Ara",
                "type": "button",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "className": "button",
                "primary": true,
                "text": "Filtrele",
                "type": "button",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "onSelect": [MockFunction],
            "orderingOtions": Array [
              Object {
                "text": "Ucuzdan Pahalıya",
                "value": "minPrice",
              },
              Object {
                "text": "Pahalıdan Ucuza",
                "value": "maxPrice",
              },
              Object {
                "text": "Tarihe Göre Yakın",
                "value": "closest",
              },
              Object {
                "text": "Tarihe Göre Uzak",
                "value": "farthest",
              },
            ],
            "selectedOrder": Object {
              "text": "Ucuzdan Pahalıya",
              "value": "minPrice",
            },
            "title": "Sıralama",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;

exports[`<OrderResults /> should match snapshot 2`] = `"<div class=\\"container container\\"><span class=\\"flightCount\\">10 Uçuş Listeleniyor</span><div class=\\"buttonsContainer\\"><button class=\\"button primary button\\" type=\\"button\\"><span class=\\"text\\">Yeniden Ara</span></button><button class=\\"button primary button\\" type=\\"button\\"><span class=\\"text\\">Filtrele</span></button></div><div class=\\"container\\"><span class=\\"title\\">Sıralama</span><div class=\\"dropdown\\"><div class=\\"selected\\"><div class=\\"selectedItem\\">Ucuzdan Pahalıya</div><div class=\\"arrow\\"></div></div></div></div></div>"`;
